### PR TITLE
v0.2.x new command interface refactor

### PR DESCRIPTION
**Pull request by nh3**
This PR adds refactored scanpy wrapper scripts featuring:

1. a subcommand-based CLI interface using click.
2. targeting scanpy-1.4.0

### CODE REVIEW

**Comment by pinin4fjords** [comment]
We need some sort of punishment device for uber-PRs such as this ;-). Will review as I can tomorrow!

**Comment by nh3** [comment]
Sorry @pinin4fjords! Initial PR is usually rather oversized. Was thinking to have more tests to make review easier, just don't have the time yet.

File: scanpy-scripts-tests.bats [hunk 1]
```diff
@@ -1,14 +1,50 @@
...
+    leiden_obj="${output_dir}/leiden.h5ad"
+    diffexp_tsv="${output_dir}/diffexp.tsv"
+    diffexp_opt="-g leiden_k10_r0_7 --reference rest --save ${diffexp_tsv}"
+    diffexp_obj="${output_dir}/diffexp.h5ad"
+}
 
-@test "Extract .mtx matrix from archive" {
-    if [ "$use_existing_outputs" = 'true' ] && [ -f "$raw_matrix" ]; then
-        skip "$raw_matrix exists and use_existing_outputs is set to 'true'"
+@test "Downlaod and extract .mtx matrix" {
```
Comment (pinin4fjords) [inline scanpy-scripts-tests.bats]: typo

File: scanpy-scripts-tests.bats [hunk 2]
```diff
@@ -1,14 +1,50 @@
 #!/usr/bin/env bats
 
 # Extract the test data
+setup() {
```
Comment (pinin4fjords) [inline scanpy-scripts-tests.bats]: Ahh, all the setup in the .bats rather than in a .sh. Neat.

File: scanpy_scripts/__init__.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+"""
+Provides version, author and exports
+"""
+
+__version__ = '0.2.0'
+
+__author__ = ', '.join([
+    'Ni Huang',
+    'Pablo Moreno',
+    'Jonathan Manning',
```
Comment (pinin4fjords) [inline scanpy_scripts/__init__.py]: Thanks for the credit! Not sure I've done enough to deserve it ;-)

**Comment by pinin4fjords** [comment]
You're forgiven @nh3 ;-). But given the complexity, it would be nice if you were a bit more verbose with the description - if only for posterity so we can look back on this PR.

File: scanpy_scripts/cmd_options.py [hunk 4]
```diff
@@ -0,0 +1,704 @@
...
+        ),
+        click.option(
+            '--output-format', '-F',
+            type=click.Choice(['anndata', 'loom', 'zarr']),
+            default='anndata',
+            show_default=True,
+            help='Output object format.',
+        ),
+        click.option(
+            '--zarr-chunk-size', '-z',
```
Comment (pinin4fjords) [inline scanpy_scripts/cmd_options.py]: We have zarr output too? Nice!

File: scanpy_scripts/cmd_options.py [hunk 5]
```diff
@@ -0,0 +1,704 @@
...
+FILTER_CMD_OPTIONS = [
+    *COMMON_OPTIONS['input'],
+    *COMMON_OPTIONS['output'],
+    click.option(
+        '--gene-name', '-g',
+        type=click.STRING,
+        default='index',
+        show_default=True,
+        help='Name of the variable that contains gene names, '
+        'used for flagging mitochondria genes.',
```
Comment (pinin4fjords) [inline scanpy_scripts/cmd_options.py]: Ahh- I've been thinking we need to do the mitochondrial thing

**Comment by nh3** [comment]
Thanks @pinin4fjords for the review! Much of the work was code-refactor to have a wrapper framework and the rest was to expose a little bit more of what scanpy is capable of. Still, only the very basic functionality is exposed so far, lacking mostly the plotting functions and those that go beyond clustering, which will be future PRs. I'll merge it for now so that people (mostly myself) can start using it and finding bugs.

**Pull request merged**